### PR TITLE
8275775: Add jcmd VM.classes to print details of all classes

### PR DESCRIPTION
**Pull request by kelthuzadx**
Add VM.classes to print details of all classes, output looks like:

1. jcmd VM.classes
```
KlassAddr Size State Flags LoaderName ClassName
0x0000000800c0b400 62 inited W bootstrap java.lang.invoke.LambdaForm$MH/0x0000000800c0b400
0x0000000800c0b000 62 inited W bootstrap java.lang.invoke.LambdaForm$DMH/0x0000000800c0b000
0x0000000800c0ac00 62 inited W bootstrap java.lang.invoke.LambdaForm$MH/0x0000000800c0ac00
...
```
2. jcmd VM.classes verbose
```
KlassAddr Size State Flags LoaderName ClassName
0x0000000800c0b400 62 inited W bootstrap java.lang.invoke.LambdaForm$MH/0x0000000800c0b400
java.lang.invoke.LambdaForm$MH/0x0000000800c0b400 {0x0000000800c0b400}
 - instance size: 2
 - klass size: 62
 - access: final synchronized
 - state: inited
 - name: 'java/lang/invoke/LambdaForm$MH+0x0000000800c0b400'
 - super: 'java/lang/Object'
 - sub:
 - arrays: NULL
 - methods: Array<T>(0x00007f620841f210)
 - method ordering: Array<T>(0x0000000800a7e5a8)
 - default_methods: Array<T>(0x0000000000000000)
 - local interfaces: Array<T>(0x00000008005af748)
 - trans. interfaces: Array<T>(0x00000008005af748)
 - constants: constant pool [41] {0x00007f620841f030} for 'java/lang/invoke/LambdaForm$MH+0x0000000800c0b400' cache=0x00007f620841f380
 - class loader data: loader data: 0x00007f61c804a690 of 'bootstrap' has a class holder
 - source file: 'LambdaForm$MH'
 - class annotations: Array<T>(0x0000000000000000)
 - class type annotations: Array<T>(0x0000000000000000)
 - field annotations: Array<T>(0x0000000000000000)
 - field type annotations: Array<T>(0x0000000000000000)
 - inner classes: Array<T>(0x00000008005af6d8)
 - nest members: Array<T>(0x00000008005af6d8)
 - permitted subclasses: Array<T>(0x00000008005af6d8)
 - java mirror: a 'java/lang/Class'{0x000000011f4b3968} = 'java/lang/invoke/LambdaForm$MH+0x0000000800c0b400'
 - vtable length 5 (start addr: 0x0000000800c0b5b8)
 - itable length 2 (start addr: 0x0000000800c0b5e0)
 - ---- static fields (1 words):
 - static final '_D_0' 'Ljava/lang/invoke/LambdaForm;' @112
 - ---- non-static fields (0 words):
 - non-static oop maps:
0x0000000800c0b000 62 inited W bootstrap java.lang.invoke.LambdaForm$DMH/0x0000000800c0b000
java.lang.invoke.LambdaForm$DMH/0x0000000800c0b000 {0x0000000800c0b000}
 - instance size: 2
 - klass size: 62
 - access: final synchronized
 - state: inited
 - name: 'java/lang/invoke/LambdaForm$DMH+0x0000000800c0b000'
 - super: 'java/lang/Object'
 - sub:
 - arrays: NULL
 - methods: Array<T>(0x00007f620841ea68)
 - method ordering: Array<T>(0x0000000800a7e5a8)
 - default_methods: Array<T>(0x0000000000000000)
 - local interfaces: Array<T>(0x00000008005af748)
 - trans. interfaces: Array<T>(0x00000008005af748)
 - constants: constant pool [49] {0x00007f620841e838} for 'java/lang/invoke/LambdaForm$DMH+0x0000000800c0b000' cache=0x00007f620841ebe0
 - class loader data: loader data: 0x00007f61c804a750 of 'bootstrap' has a class holder
 - source file: 'LambdaForm$DMH'
 - class annotations: Array<T>(0x0000000000000000)
 - class type annotations: Array<T>(0x0000000000000000)
 - field annotations: Array<T>(0x0000000000000000)
 - field type annotations: Array<T>(0x0000000000000000)
 - inner classes: Array<T>(0x00000008005af6d8)
 - nest members: Array<T>(0x00000008005af6d8)
 - permitted subclasses: Array<T>(0x00000008005af6d8)
 - java mirror: a 'java/lang/Class'{0x000000011f4b0968} = 'java/lang/invoke/LambdaForm$DMH+0x0000000800c0b000'
 - vtable length 5 (start addr: 0x0000000800c0b1b8)
 - itable length 2 (start addr: 0x0000000800c0b1e0)
 - ---- static fields (1 words):
 - static final '_D_0' 'Ljava/lang/invoke/LambdaForm;' @112
 - ---- non-static fields (0 words):
...
```

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8275775](https://bugs.openjdk.java.net/browse/JDK-8275775): Add jcmd VM.classes to print details of all classes

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**)
 * [Thomas Stuefe](https://openjdk.java.net/census#stuefe) (@tstuefe - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7105/head:pull/7105` \
`$ git checkout pull/7105`

Update a local copy of the PR: \
`$ git checkout pull/7105` \
`$ git pull https://git.openjdk.java.net/jdk pull/7105/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7105`

View PR using the GUI difftool: \
`$ git pr show -t 7105`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7105.diff">https://git.openjdk.java.net/jdk/pull/7105.diff</a>

</details>

### CODE REVIEW

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Hi,
This seems reasonable in general but a few details to work through - see comments below.
Thanks,
David

**Comment by plummercj** [comment]
It seems it would be useful to support the verbose output with just a single class that is specified, although that would suggest that the dcmd name should then be something other than `VM.classes`.

File: src/hotspot/share/services/diagnosticCommand.hpp
Comment (kelthuzadx) [reply src/hotspot/share/services/diagnosticCommand.hpp:873]: Thanks for reviews! Since ClassHierarchyDCmd uses `"Medium: Depends on number of loaded classes."`, so I'm going to change the impact description but keeping as `Medium` level.

File: src/hotspot/share/oops/instanceKlass.cpp
Comment (kelthuzadx) [reply src/hotspot/share/oops/instanceKlass.cpp:2103]: These methods belong to `Klass`

File: src/hotspot/share/oops/instanceKlass.cpp
Comment (kelthuzadx) [reply src/hotspot/share/oops/instanceKlass.cpp:2100]: I don't find a proper place to document these flags, do you have any suggestions?

I do think we can output flag explanations as well, but it looks somewhat strange..
```
Flags: V=..., W=...
KlassAddr           Size  State                 Flags    LoaderName    ClassName  
0x0000000800c0b400  62    fully_initialized     W        bootstrap     java.lang.invoke.LambdaForm$MH/0x0000000800c0b400  
0x0000000800c0b000  62    fully_initialized     W        bootstrap     java.lang.invoke.LambdaForm$DMH/0x0000000800c0b000  
0x0000000800c0ac00  62    fully_initialized     W        bootstrap     java.lang.invoke.LambdaForm$MH/0x0000000800c0ac00  
0x0000000800c0a800  62    fully_initialized     W        bootstrap     java.lang.invoke.LambdaForm$MH/0x0000000800c0a800  
0x0000000800c0a400  62    fully_initialized     W        bootstrap     java.lang.invoke.LambdaForm$MH/0x0000000800c0a400 
```

**Comment by kelthuzadx** [comment]
> It seems it would be useful to support the verbose output with just a single class that is specified, although that would suggest that the dcmd name should then be something other than `VM.classes`.

This is a good idea, but `jcmd VM.classes verbose=XX` looks strange, `jcmd VM.class XX` is also not much proper, because we desire to print all classes in default(`jcmd VM.class`). an alternative is to use `jcmd VM.classes verbose | grep XX` currently.

**Comment by plummercj** [comment]
> > It seems it would be useful to support the verbose output with just a single class that is specified, although that would suggest that the dcmd name should then be something other than `VM.classes`.
> 
> This is a good idea, but `jcmd VM.classes verbose=XX` looks strange, `jcmd VM.class XX` is also not much proper, because we desire to print all classes in default(`jcmd VM.class`). an alternative is to use `jcmd VM.classes verbose | grep XX` currently.

I was thinking the syntax would look like:  `jcmd VM.classes [verbose [classname]]`

Your grep solution doesn't work because each class has multiple lines of output.

File: src/hotspot/share/oops/instanceKlass.cpp
Comment (dholmes-ora) [reply src/hotspot/share/oops/instanceKlass.cpp:2095]: That is true (though I wonder if it should be) but the question remains can these ever be present on a non-instance class?

Somewhat related can you show the output for an array class please.

File: src/hotspot/share/oops/instanceKlass.cpp
Comment (dholmes-ora) [reply src/hotspot/share/oops/instanceKlass.cpp:2092]: Printing a legend line would be good, but there should also be actual documentation in the help output I think.

**Comment by iklam** [comment]
> > > It seems it would be useful to support the verbose output with just a single class that is specified, although that would suggest that the dcmd name should then be something other than `VM.classes`.
> > 
> > 
> > This is a good idea, but `jcmd VM.classes verbose=XX` looks strange, `jcmd VM.class XX` is also not much proper, because we desire to print all classes in default(`jcmd VM.class`). an alternative is to use `jcmd VM.classes verbose | grep XX` currently.
> 
> I was thinking the syntax would look like: `jcmd VM.classes [verbose [classname]]`
> 
> Your grep solution doesn't work because each class has multiple lines of output.

How about this:

```
jcmd VM.classes -verbose classname classname ...
```

-verbose is optional

more than one classnames can be specified.

if no classnames are specified, all classes are printed

**Comment by navyxliu** [comment]
> > > > It seems it would be useful to support the verbose output with just a single class that is specified, although that would suggest that the dcmd name should then be something other than `VM.classes`.
> > > 
> > > 
> > > This is a good idea, but `jcmd VM.classes verbose=XX` looks strange, `jcmd VM.class XX` is also not much proper, because we desire to print all classes in default(`jcmd VM.class`). an alternative is to use `jcmd VM.classes verbose | grep XX` currently.
> > 
> > 
> > I was thinking the syntax would look like: `jcmd VM.classes [verbose [classname]]`
> > Your grep solution doesn't work because each class has multiple lines of output.
> 
> How about this:
> 
> ```
> jcmd VM.classes -verbose classname classname ...
> ```
> 
> -verbose is optional
> 
> more than one classnames can be specified.
> 
> if no classnames are specified, all classes are printed
If the class name here means the "fully-qualified" class name, I guess it's not practical to input multiple classnames like "java.lang.invoke.LambdaForm$MH/0x0000000800c0b400" in cmdline. 

The main cost of VM_PrintClasses should be the traversal of the all classes. I feel a filter won't save much runtime time.  We can leave it to the external awk scripts. What do you think?

**Comment by iklam** [comment]
> > How about this:
> > ```
> > jcmd VM.classes -verbose classname classname ...
> > ```
> >   
> > -verbose is optional
> > more than one classnames can be specified.
> > if no classnames are specified, all classes are printed
> 
> If the class name here means the "fully-qualified" class name, I guess it's not practical to input multiple classnames like "java.lang.invoke.LambdaForm$MH/0x0000000800c0b400" in cmdline.
> 
> The main cost of VM_PrintClasses should be the traversal of all classes. I feel a filter won't save much runtime time. We can leave it to the external awk scripts. What do you think?

That sounds fair. I think for filtering it's best left to external tools.

I think we should use `-verbose` as the optional argument, to be consistent with other jcmds such as ` VM.symboltable`

**Comment by kelthuzadx** [comment]
Thank you all for suggestions and feedback. I add the help document for flags and use `-verbose` to print detailed content of classes, to be consistent with other jcmd commands.

File: src/hotspot/share/oops/instanceKlass.cpp
Comment (iklam) [inline src/hotspot/share/oops/instanceKlass.cpp:2106]: For custom class loaders, this will likely print a long class name that will over the 12 character limit, making the output somewhat hard to read.


```
const char* ClassLoaderData::loader_name() const {
   if (_class_loader_klass == NULL) {
     return BOOTSTRAP_LOADER_NAME;
   } else if (_name != NULL) {
     return _name->as_C_string();
   } else {
     return _class_loader_klass->external_name();
   }
}
```

Also, for custom loaders, printing out just the name of the loader class is not sufficient, as multiple loader instances may have the same type.

Maybe we should just remove line 2106? If the user wants to know the class loader, they can use the "-verbose" option of this jcmd.

File: src/hotspot/share/services/diagnosticCommand.hpp
Comment (iklam) [inline src/hotspot/share/services/diagnosticCommand.hpp:870]: I think it's better to say "Print all loaded classes". Most commands in this file do not use the third-person singular verb ending. The words "list of" are redundant.

File: test/hotspot/jtreg/runtime/CommandLine/PrintClasses.java
Comment (iklam) [inline test/hotspot/jtreg/runtime/CommandLine/PrintClasses.java:40]: For sanity, I think we should have two test cases, one with -verbose, and one without.

**Review by iklam — COMMENTED** [review]
This looks generally OK to me. Some minor suggestions.

File: src/hotspot/share/oops/instanceKlass.cpp
Comment (kelthuzadx) [reply src/hotspot/share/oops/instanceKlass.cpp]: Thanks for review! All comments are addressed.

File: src/hotspot/share/oops/instanceKlass.cpp
Comment (iklam) [inline src/hotspot/share/oops/instanceKlass.cpp:2081]: Should be `%4d` so that the numbers are aligned correctly.

**Review by iklam — APPROVED** [review]
LGTM. One minor nit,

Could you update the PR description to include examples of the final output format.

File: src/hotspot/share/oops/instanceKlass.cpp
Comment (kelthuzadx) [reply src/hotspot/share/oops/instanceKlass.cpp]: @dholmes-ora David, Can you please take a look at the latest version? I've addressed all problems you pointed out.

File: src/hotspot/share/services/diagnosticCommand.cpp
Comment (dholmes-ora) [inline src/hotspot/share/services/diagnosticCommand.cpp:964]: typo finializer - but should be finalize

Is this actually only present for "non-trivial finalize" method?

**Review by dholmes-ora — COMMENTED** [review]
Hi Yi,

I had been expecting to see further updates as not all issues seem resolved. I have a few further typos and nits below. But I'd like to see someone from serviceability actually approve this.

Thanks,
David

File: src/hotspot/share/oops/instanceKlass.cpp
Comment (kelthuzadx) [reply src/hotspot/share/oops/instanceKlass.cpp:2081]: Current:
```
$./jcmd 83908 VM.classes|head -10
83908:
KlassAddr           Size  State                 Flags    ClassName  
0x0000000800df8400  62    fully_initialized     W        java.lang.invoke.LambdaForm$DMH/0x0000000800df8400  
0x0000000800df8000  62    fully_initialized     W        java.lang.invoke.LambdaForm$DMH/0x0000000800df8000  
0x0000000800de4400  62    fully_initialized     W        java.lang.invoke.LambdaForm$DMH/0x0000000800de4400  
0x0000000800de4000  62    fully_initialized     W        java.lang.invoke.LambdaForm$DMH/0x0000000800de4000  
0x0000000800dc8800  62    fully_initialized     W        java.lang.invoke.LambdaForm$DMH/0x0000000800dc8800  
0x0000000800dc8400  62    fully_initialized     W        java.lang.invoke.LambdaForm$DMH/0x0000000800dc8400  
0x0000000800dc8000  62    fully_initialized     W        java.lang.invoke.LambdaForm$DMH/0x0000000800dc8000  
0x0000000800db9800  62    fully_initialized     W        java.lang.invoke.LambdaForm$DMH/0x0000000800db9800 
```
After using "%-4d":
```
$./jcmd 75481 VM.classes|head
75481:
KlassAddr           Size  State                 Flags    ClassName  
0x0000000800df8400    62  fully_initialized     W        java.lang.invoke.LambdaForm$DMH/0x0000000800df8400  
0x0000000800df8000    62  fully_initialized     W        java.lang.invoke.LambdaForm$DMH/0x0000000800df8000  
0x0000000800de4400    62  fully_initialized     W        java.lang.invoke.LambdaForm$DMH/0x0000000800de4400  
0x0000000800de4000    62  fully_initialized     W        java.lang.invoke.LambdaForm$DMH/0x0000000800de4000 
```
So we do not need to change this.

File: src/hotspot/share/services/diagnosticCommand.cpp
Comment (kelthuzadx) [reply src/hotspot/share/services/diagnosticCommand.cpp]: I'm not sure what's the meaning of "non-trivial finalize" method, can you elaborate more for it?

**Comment by kelthuzadx** [comment]
@plummercj Can you please help review this from the serviceability point of view? Thanks in advance!

File: src/hotspot/share/services/diagnosticCommand.cpp
Comment (dholmes-ora) [inline src/hotspot/share/services/diagnosticCommand.cpp:964]: This is still spelt incorrectly: finalize

**Review by dholmes-ora — CHANGES_REQUESTED** [review]

File: src/hotspot/share/services/diagnosticCommand.cpp
Comment (kelthuzadx) [reply src/hotspot/share/services/diagnosticCommand.cpp]: Sorry... changed.

File: src/hotspot/share/services/diagnosticCommand.cpp
Comment (dholmes-ora) [reply src/hotspot/share/services/diagnosticCommand.cpp]: I mean a finalize() method that actually does something. I checked the code and you will print F is the current class has a non-empty finalize() method, or it has a superclass with a non-empty finalize method. I would suggest updating the text to:

`F = has, or inherits, a non-empty finalize method`

Thanks,
David

Comment (kelthuzadx) [reply src/hotspot/share/services/diagnosticCommand.cpp]: Done. This description is more clear compared to non-trivial finalize.

**Review by dholmes-ora — APPROVED** [review]

File: src/hotspot/share/oops/instanceKlass.cpp
Comment (iklam) [reply src/hotspot/share/oops/instanceKlass.cpp:2093]: You should change it to `%4d`. Otherwise, when the numbers are changed in the future (e.g., to 3 or 4 digits) they will be misaligned:

```
KlassAddr           Size  State                 Flags    ClassName  
0x0000000800df8400  62    fully_initialized     W        java.lang.invoke.LambdaForm$DMH/0x0000000800df8400  
0x0000000800df8000  123   fully_initialized     W        java.lang.invoke.LambdaForm$DMH/0x0000000800df8000  
0x0000000800de4400  4567  fully_initialized     W        java.lang.invoke.LambdaForm$DMH/0x0000000800de4400  
```

**Comment by linzang** [comment]
Sorry that I just chime in. It seems this change adds new command options, so it seems that `csr` is required?

**Comment by kelthuzadx** [comment]
> Sorry that I just chime in. It seems this change adds new command options, so it seems that `csr` is required?

Hi @linzang, according to [previous discussion](https://github.com/openjdk/jdk/pull/6075) and [comments in JBS](https://bugs.openjdk.java.net/browse/JDK-8275775), it's not necessary to create a csr for it.

File: src/hotspot/share/oops/instanceKlass.cpp
Comment (kelthuzadx) [reply src/hotspot/share/oops/instanceKlass.cpp:2093]: > You should change it to `%4d`. Otherwise, when the numbers are changed in the future (e.g., to 3 or 4 digits) they will be misaligned:
> 
> ```
> KlassAddr           Size  State                 Flags    ClassName  
> 0x0000000800df8400  62    fully_initialized     W        java.lang.invoke.LambdaForm$DMH/0x0000000800df8400  
> 0x0000000800df8000  123   fully_initialized     W        java.lang.invoke.LambdaForm$DMH/0x0000000800df8000  
> 0x0000000800de4400  4567  fully_initialized     W        java.lang.invoke.LambdaForm$DMH/0x0000000800de4400  
> ```

This format looks pretty good to me, they are all aligned to left. If you still think it's more proper to have a format like this:
```
KlassAddr           Size  State                 Flags    ClassName  
0x0000000800df8400  62    fully_initialized     W        java.lang.invoke.LambdaForm$DMH/0x0000000800df8400  
0x0000000800df8000  123   fully_initialized     W        java.lang.invoke.LambdaForm$DMH/0x0000000800df8000  
0x0000000800de4400  4567  fully_initialized     W        java.lang.invoke.LambdaForm$DMH/0x0000000800de4400  
```
Then I'm glad to do so ;)

**Comment by linzang** [comment]
> > Sorry that I just chime in. It seems this change adds new command options, so it seems that `csr` is required?
> 
> Hi @linzang, according to [previous discussion](https://github.com/openjdk/jdk/pull/6075) and [comments in JBS](https://bugs.openjdk.java.net/browse/JDK-8275775), it's not necessary to create a csr for it.

Ah, I missed that. thanks for point it out!

File: src/hotspot/share/oops/instanceKlass.cpp
Comment (iklam) [reply src/hotspot/share/oops/instanceKlass.cpp:2093]: Numbers should be aligned to the right. The following is what I want:

```
  62
 123
4567
```

Comment (kelthuzadx) [reply src/hotspot/share/oops/instanceKlass.cpp:2093]: > Numbers should be aligned to the right. The following is what I want:
> 
> ```
>   62
>  123
> 4567
> ```

Done.

**Review by iklam — APPROVED** [review]
LGTM

**Review by dholmes-ora — APPROVED** [review]

**Review by tstuefe — APPROVED** [review]
Looks good!

**Comment by kelthuzadx** [comment]
Thank you for taking time to help review. 3 approval, I want to merge this tomorrow if no objections/comments.

**Comment by kelthuzadx** [comment]
> _Mailing list message from [Bernd Eckenfels](mailto:ecki@zusammenkunft.net) on [hotspot-dev](mailto:hotspot-dev@mail.openjdk.java.net):_
> 
> Hello,
> 
> I would add an additional argument to allow substring filtering on the fully qualified class name (like com/example or UtilClass), since this can creator reduce processing/printing time. But I guess that can be added as an additional feature later on (maybe only the ?verbose? variant can conflict if this is not a option (-verbose?).
> 
> Gruss Bernd -- http://bernd.eckenfels.net ________________________________ Von: serviceability-dev <serviceability-dev-retn at openjdk.java.net> im Auftrag von Yi Yang <yyang at openjdk.java.net> Gesendet: Monday, March 7, 2022 3:33:59 AM An: hotspot-dev at openjdk.java.net <hotspot-dev at openjdk.java.net>; serviceability-dev at openjdk.java.net <serviceability-dev at openjdk.java.net> Betreff: Re: RFR: 8275775: Add jcmd VM.classes to print details of all classes [v10]
> 
> On Fri, 4 Mar 2022 09:05:36 GMT, Yi Yang <yyang at openjdk.org> wrote:
> 
> > > Add VM.classes to print details of all classes, output looks like:
> > > 1. jcmd VM.classes
> > > KlassAddr Size State Flags LoaderName ClassName
> > > 0x0000000800c0b400 62 inited W bootstrap java.lang.invoke.LambdaForm$MH/0x0000000800c0b400
> > > 0x0000000800c0b000 62 inited W bootstrap java.lang.invoke.LambdaForm$DMH/0x0000000800c0b000
> > > 0x0000000800c0ac00 62 inited W bootstrap java.lang.invoke.LambdaForm$MH/0x0000000800c0ac00
> > > ...
> > > 2. jcmd VM.classes verbose
> > > KlassAddr Size State Flags LoaderName ClassName
> > > 0x0000000800c0b400 62 inited W bootstrap java.lang.invoke.LambdaForm$MH/0x0000000800c0b400
> > > java.lang.invoke.LambdaForm$MH/0x0000000800c0b400 {0x0000000800c0b400}
> > > - instance size: 2
> > > - klass size: 62
> > > - access: final synchronized
> > > - state: inited
> > > - name: 'java/lang/invoke/LambdaForm$MH+0x0000000800c0b400'
> > > - super: 'java/lang/Object'
> > > - sub:
> > > - arrays: NULL
> > > - methods: Array<T>(0x00007f620841f210)
> > > - method ordering: Array<T>(0x0000000800a7e5a8)
> > > - default_methods: Array<T>(0x0000000000000000)
> > > - local interfaces: Array<T>(0x00000008005af748)
> > > - trans. interfaces: Array<T>(0x00000008005af748)
> > > - constants: constant pool [41] {0x00007f620841f030} for 'java/lang/invoke/LambdaForm$MH+0x0000000800c0b400' cache=0x00007f620841f380
> > > - class loader data: loader data: 0x00007f61c804a690 of 'bootstrap' has a class holder
> > > - source file: 'LambdaForm$MH'
> > > - class annotations: Array<T>(0x0000000000000000)
> > > - class type annotations: Array<T>(0x0000000000000000)
> > > - field annotations: Array<T>(0x0000000000000000)
> > > - field type annotations: Array<T>(0x0000000000000000)
> > > - inner classes: Array<T>(0x00000008005af6d8)
> > > - nest members: Array<T>(0x00000008005af6d8)
> > > - permitted subclasses: Array<T>(0x00000008005af6d8)
> > > - java mirror: a 'java/lang/Class'{0x000000011f4b3968} = 'java/lang/invoke/LambdaForm$MH+0x0000000800c0b400'
> > > - vtable length 5 (start addr: 0x0000000800c0b5b8)
> > > - itable length 2 (start addr: 0x0000000800c0b5e0)
> > > - ---- static fields (1 words):
> > > - static final '_D_0' 'Ljava/lang/invoke/LambdaForm;' @112
> > > - ---- non-static fields (0 words):
> > > - non-static oop maps:
> > > 0x0000000800c0b000 62 inited W bootstrap java.lang.invoke.LambdaForm$DMH/0x0000000800c0b000
> > > java.lang.invoke.LambdaForm$DMH/0x0000000800c0b000 {0x0000000800c0b000}
> > > - instance size: 2
> > > - klass size: 62
> > > - access: final synchronized
> > > - state: inited
> > > - name: 'java/lang/invoke/LambdaForm$DMH+0x0000000800c0b000'
> > > - super: 'java/lang/Object'
> > > - sub:
> > > - arrays: NULL
> > > - methods: Array<T>(0x00007f620841ea68)
> > > - method ordering: Array<T>(0x0000000800a7e5a8)
> > > - default_methods: Array<T>(0x0000000000000000)
> > > - local interfaces: Array<T>(0x00000008005af748)
> > > - trans. interfaces: Array<T>(0x00000008005af748)
> > > - constants: constant pool [49] {0x00007f620841e838} for 'java/lang/invoke/LambdaForm$DMH+0x0000000800c0b000' cache=0x00007f620841ebe0
> > > - class loader data: loader data: 0x00007f61c804a750 of 'bootstrap' has a class holder
> > > - source file: 'LambdaForm$DMH'
> > > - class annotations: Array<T>(0x0000000000000000)
> > > - class type annotations: Array<T>(0x0000000000000000)
> > > - field annotations: Array<T>(0x0000000000000000)
> > > - field type annotations: Array<T>(0x0000000000000000)
> > > - inner classes: Array<T>(0x00000008005af6d8)
> > > - nest members: Array<T>(0x00000008005af6d8)
> > > - permitted subclasses: Array<T>(0x00000008005af6d8)
> > > - java mirror: a 'java/lang/Class'{0x000000011f4b0968} = 'java/lang/invoke/LambdaForm$DMH+0x0000000800c0b000'
> > > - vtable length 5 (start addr: 0x0000000800c0b1b8)
> > > - itable length 2 (start addr: 0x0000000800c0b1e0)
> > > - ---- static fields (1 words):
> > > - static final '_D_0' 'Ljava/lang/invoke/LambdaForm;' @112
> > > - ---- non-static fields (0 words):
> > > ...
> > 
> > 
> > Yi Yang has updated the pull request incrementally with one additional commit since the last revision:
> > use %4d
> 
> Thank you for taking time to help review. 3 approval, I want to merge this tomorrow if no objections/comments.
> 
> -------------
> 
> PR: https://git.openjdk.java.net/jdk/pull/7105

Hi Bernd,

Class filter has been discussed before on PR(Please follow previous comments for more details). The conclusion is a filter won't save much runtime time. We can leave it to the external tools(grep/awk,etc).

**Pull request closed without merging**
